src/collator.rs

```diff
@@ -97,31 +102,46 @@ impl<F: crate::system_snapshot::SystemSnapshotter> Collator<F> {
         facts.append(&mut correlation_data.properties);
         groups.append(&mut correlation_data.groups_as_map());
 
+        let stored_ident = storage.load().await.ok().flatten();
+
         Self {
-            system_snapshotter,
+            system_snapshotter: Mutex::new(system_snapshotter),
+            storage,
             incoming,
             outgoing,
             session_id: correlation_data
                 .session_id
                 .unwrap_or_else(|| uuid::Uuid::now_v7().to_string()),
-            anon_distinct_id: correlation_data
-                .anon_distinct_id
-                .unwrap_or_else(|| uuid::Uuid::now_v7().to_string()),
-            distinct_id: distinct_id.or(correlation_data.distinct_id),
-            device_id: device_id.or(correlation_data.device_id).unwrap_or_default(),
+            anon_distinct_id: anonymous_distinct_id
+                .or(stored_ident
+                    .as_ref()
+                    .map(|props| props.anonymous_distinct_id.clone()))
+                .or(correlation_data
+                    .anon_distinct_id
+                    .map(AnonymousDistinctId::from))
+                .unwrap_or_else(|| AnonymousDistinctId::from(uuid::Uuid::now_v7().to_string())),
```

Comment:
I think this (and other `.or` calls here) were meant to be lazy `.or_else` calls:
```suggestion
                .or_else(|| {
                    stored_ident
                        .as_ref()
                        .map(|props| props.anonymous_distinct_id.clone())
                })
                .or_else(|| {
                    correlation_data
                        .anon_distinct_id
                        .map(AnonymousDistinctId::from)
                })
                .unwrap_or_else(|| AnonymousDistinctId::from(uuid::Uuid::now_v7().to_string())),

```